File: Project Euler/p_35.py
import math

max_size = 1000000
primes = [True] * max_size

for i in range(4, max_size, 2):
    primes[i] = False

r = int(math.sqrt(max_size))

for i in range(3, r, 2):
    if (primes[i] == True):
        for j in range(i * i, max_size, i + i):
            primes[j] = False

def circle(n):
    nums = str(n)
    nums = nums + nums[0]
    nums = nums[1:]
    
    return int(nums)

def is_circular(n):
    temp = n
        
    temp = circle(temp)
    
    while (temp != n):
        if primes[temp] == False:
            return False
        else:
            temp = circle(temp)
    
    return True

cnt = 0

for i in range(2, 100):
    if (primes[i] == True):
        if is_circular(i):
            cnt += 1
            
print(cnt)

File: Project Euler/p_35.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define max_size 1000000
bool *primes;

void sieve()
{
    int i, j;

    primes = (bool *) calloc(max_size, sizeof(bool));

    for (i = 4; i < max_size; i += 2) {
        primes[i] = true;
    }

    for (i = 3; i * i < max_size; i += 2) {
        if (primes[i] == false) {
            for (j = i * i; j < max_size; j += i + i) {
                primes[j] = true;
            }
        }
    }
}

void circle(int *n)
{
    int temp = *n, len = 10;

    while (temp / 10 > 0) {
        temp /= 10;
        len *= 10;
    }

    *n = (*n * 10 + temp) % len;
}

int is_circular(int n)
{
    int temp = n;

    if (primes[temp] == false) {
        circle(&temp);

        if (primes[temp] == true) {
            return false;
        }

        while (n != temp) {
            if (primes[temp] == true) {
                return false;
            }
            else {
                circle(&temp);
            }
        }

        return true;
    }
    else {
        return false;
    }
}

int main()
{
    sieve();
    int i, j, temp, count = 0;

    for (i = 2; i < 100; i++) {
        temp = i;

        if (is_circular(i) == true) {
            // printf("%d\n", i);

            count++;
        }
    }

    printf("%d\n", count);

    return 0;
}